driver-core/src/main/java/com/datastax/driver/core/FrameCompressor.java

```diff
@@ -58,23 +60,119 @@ private SnappyCompressor() {
+                    // Increase reader index.
+                    input.readerIndex(input.writerIndex());
+
+                    ByteBuffer out = output.internalNioBuffer(output.writerIndex(), output.writableBytes());
+                    int written = Snappy.compress(in, out);
+                    // Set the writer index so the amount of written bytes is reflected
+                    output.writerIndex(output.writerIndex() + written);
+                    frameBody = output;
+                } catch (IOException e) {
+                    // release output buffer so we not leak and rethrow exception.
+                    output.release();
+                    throw e;
+                }
+            } else {
+                int inOffset = input.arrayOffset() + input.readerIndex();
+                byte[] in = input.array();
+                int len = input.readableBytes();
+                // Increase reader index.
+                input.readerIndex(input.writerIndex());
+
+                // Allocate a heap buffer from the ByteBufAllocator as we may use a PooledByteBufAllocator and so
+                // can eliminate the overhead of allocate a new byte[].
+                ByteBuf output = input.alloc().heapBuffer(maxCompressedLength);
+                try {
+                    // Calculate the correct offset.
+                    int offset = output.arrayOffset() + output.writerIndex();
+                    byte[] out = output.array();
+                    int written = Snappy.compress(in, inOffset, len, out, offset);
+
+                    // Increase the writerIndex with the written bytes.
+                    output.writerIndex(output.writerIndex() + written);
+                    frameBody = output;
+                } catch (IOException e) {
+                    // release output buffer so we not leak and rethrow exception.
+                    output.release();
+                    throw e;
+                }
+            }
+            return frame.with(frameBody);
         }
 
         @Override
         public Frame decompress(Frame frame) throws IOException {
-            byte[] input = CBUtil.readRawBytes(frame.body);
-
-            if (!Snappy.isValidCompressedBuffer(input, 0, input.length))
-                throw new DriverInternalError("Provided frame does not appear to be Snappy compressed");
-
-            byte[] output = new byte[Snappy.uncompressedLength(input)];
-            int size = Snappy.uncompress(input, 0, input.length, output, 0);
-            return frame.with(Unpooled.wrappedBuffer(output, 0, size));
+            ByteBuf input = frame.body;
+            final ByteBuf frameBody;
+
+            if (input.isDirect()) {
+                // Using internalNioBuffer(...) as we only hold the reference in this method and so can
+                // reduce Object allocations.
+                ByteBuffer in = input.internalNioBuffer(input.readerIndex(), input.readableBytes());
```

Comment:
Just noticed on #726 that for the input buffer you made a convenience method that was used checked the nioBufferCount on the input buffer, guessing we need this here too?

```java
private static ByteBuffer inputNioBuffer(ByteBuf buf, int index, int len) {
    // Using internalNioBuffer(...) as we only hold the reference in this method and so can
    // reduce Object allocations.
    return buf.nioBufferCount() == 1 ? buf.internalNioBuffer(index, len) : buf.nioBuffer(index, len);
}
```